File: plan-dashboard/scripts/lib/para-blocks.mjs
/**
 * para-blocks.mjs — the prose blocks of a chapter, and a stable name for each.
 *
 * JAVASCRIPT HALF OF A MIRROR PAIR. The Python half is
 * `scripts/podcast/_para_blocks.py`. Both are pinned to the SAME fixtures at
 * `para-blocks.fixtures.json`, so a change to either that is not matched in the
 * other fails a test rather than letting the aligner and the Composer disagree
 * about which blocks a chapter HAS or what each one is called. A silent
 * disagreement here does not merely degrade the Arabic reveal — it puts the wrong
 * Arabic above the wrong paragraph, which is worse than showing none.
 *
 * WHY FINGERPRINT THE RAW MARKDOWN, not the rendered text: both sides read the
 * same `book/book.md` body, so the raw block is a shared, byte-identical input.
 * Fingerprinting rendered text would drag in `passage-match.foldText`, which
 * exists to reconcile two DIFFERENT renderings of one sentence and is a far
 * heavier fold — a real drift risk to mirror, for no gain here.
 */
import { createHash } from "node:crypto";

/**
 * A block is PROSE unless it opens as a blockquote, heading, or raw HTML block.
 * This is the rule `lib/reader/composer.ts` already applied inline for its
 * `paras` count, which in turn mirrors what `visual-layout.mjs` counts as a
 * paragraph when placing figures — so all three agree on what "paragraph 4 of
 * this chapter" means. Lifted here so there is one copy, not three.
 */
const NOT_PROSE_RE = /^\s*[>#<]/;

/**
 * A block that is NOTHING BUT a standalone `![alt](src)` line — the same
 * "whole line, nothing else" test markdown.ts's own image branch uses to
 * decide a figure rather than literal text. Added 2026-08-14 alongside the
 * ChapterImage editor node: an image line renders as a `chapterImage` node in
 * the live doc, a type `alignablePositions` (book-composer.ts) deliberately
 * does not count as an alignable paragraph — but this function, reading the
 * SAME markdown independently, was still counting that line as a numbered
 * prose block. The two counts then disagreed by one for every image in the
 * chapter, `alignablePositions` saw `found.length !== keys.length`, and every
 * alignment button in the whole chapter disabled itself rather than risk
 * pointing at the wrong paragraph — found live 2026-08-14 in a chapter with
 * four images, where the align buttons never enabled at all. An image mid-
 * sentence is UNCHANGED by this: it stays literal text in a real paragraph on
 * both sides, exactly as it did before ChapterImage existed.
 */
const IMAGE_ONLY_RE = /^!\[[^\]]*\]\([^)\s]+\)$/;

/** The chapter's prose blocks, in order, trimmed. */
export function proseBlocks(body) {
  return String(body ?? "")
    .split(/\n\s*\n/)
    .map((b) => b.trim())
    .filter((b) => b && !NOT_PROSE_RE.test(b) && !IMAGE_ONLY_RE.test(b));
}

/**
 * A stable short name for one prose block.
 *
 * Whitespace-collapsed and lowercased before hashing, so a re-wrap or a change of
 * indentation does not rename a paragraph a reader would call the same one.
 * Anything more aggressive would start merging paragraphs that genuinely differ.
 */
export function paraFingerprint(block) {
  const norm = String(block ?? "")
    .replace(/\s+/g, " ")
    .trim()
    .toLowerCase();
  return createHash("sha256").update(norm, "utf8").digest("hex").slice(0, 16);
}

/** `paraFingerprint` of every prose block, in order. */
export function fingerprints(body) {
  return proseBlocks(body).map(paraFingerprint);
}

/**
 * One name for a chapter's WHOLE ordered block list — the incremental key. When
 * this is unchanged the chapter's prose has not moved, so a stored alignment is
 * still valid and re-composing costs nothing.
 */
export function blocksFingerprint(body) {
  return createHash("sha256")
    .update(fingerprints(body).join("\n"), "utf8")
    .digest("hex")
    .slice(0, 16);
}

File: scripts/podcast/_para_blocks.py
"""_para_blocks.py — the prose blocks of a chapter, and a stable name for each.

PYTHON HALF OF A MIRROR PAIR. The JavaScript half is
``plan-dashboard/scripts/lib/para-blocks.mjs``. Both are pinned to the SAME
fixtures at ``plan-dashboard/scripts/lib/para-blocks.fixtures.json``, so a change
to either that is not matched in the other fails a test rather than letting the
aligner and the Composer disagree about which blocks a chapter HAS or what each
one is called. A silent disagreement here does not degrade the Arabic reveal — it
puts the wrong Arabic above the wrong paragraph, which is worse than showing none.

WHY FINGERPRINT THE RAW MARKDOWN, and not the rendered text. Both sides read the
same `book/book.md` body, so the raw block is a shared, byte-identical input.
Fingerprinting rendered text would drag in `passage-match.foldText`, which exists
to reconcile two DIFFERENT renderings of one sentence (the Composer keeps
scholarly transliteration where the live reader folds it) and is a much heavier
NFD/combining-mark fold — a real drift risk to port for no gain here.

WHY NOT A POSITIONAL INDEX. Composition is not stable: articulation turned 61
source paragraphs into 133 in one chapter, and any Composer edit that splits a
paragraph shifts every index after it. An index would then place confidently
wrong Arabic above everything downstream, with nothing to signal it. A
fingerprint fails CLOSED — a rewritten paragraph simply has no entry, and the
reveal falls back to the bracket its neighbours pin.
"""

from __future__ import annotations

import hashlib
import re

# A chapter body is split on blank lines; a block is PROSE unless it opens as a
# blockquote, a heading, or a raw HTML block. This mirrors the `paras` count in
# `plan-dashboard/src/lib/reader/composer.ts`, which in turn mirrors what
# `visual-layout.mjs` counts as a paragraph when it places figures — so the three
# agree on what "paragraph 4 of this chapter" means.
_NOT_PROSE_RE = re.compile(r"^\s*[>#<]")
_SPLIT_RE = re.compile(r"\n\s*\n")

# A block that is NOTHING BUT a standalone `![alt](src)` line — see
# para-blocks.mjs's own docstring on this constant for the full history: an
# image line renders as a `chapterImage` editor node the Composer's alignment
# system deliberately does not count as an alignable paragraph, so counting it
# here too disagreed with that count by one per image and disabled every
# alignment button in the chapter.
_IMAGE_ONLY_RE = re.compile(r"^!\[[^\]]*\]\([^)\s]+\)$")


def prose_blocks(body: str) -> list[str]:
    """The chapter's prose blocks, in order, stripped."""
    out = []
    for block in _SPLIT_RE.split(body or ""):
        stripped = block.strip()
        if not stripped:
            continue
        if _NOT_PROSE_RE.match(block):
            continue
        if _IMAGE_ONLY_RE.match(stripped):
            continue
        out.append(stripped)
    return out


def para_fingerprint(block: str) -> str:
    """A stable short name for one prose block.

    Whitespace-collapsed and lowercased before hashing, so a re-wrap or a change
    of indentation does not rename a paragraph that a reader would call the same
    one. Anything more aggressive would start merging paragraphs that genuinely
    differ.
    """
    norm = re.sub(r"\s+", " ", block or "").strip().lower()
    return hashlib.sha256(norm.encode("utf-8")).hexdigest()[:16]


def fingerprints(body: str) -> list[str]:
    """`para_fingerprint` of every prose block, in order."""
    return [para_fingerprint(b) for b in prose_blocks(body)]


def blocks_fingerprint(body: str) -> str:
    """One name for a chapter's WHOLE ordered block list.

    The incremental key: when this is unchanged the chapter's prose has not moved,
    so its stored alignment is still valid and re-composing costs nothing.
    """
    joined = "\n".join(fingerprints(body))
    return hashlib.sha256(joined.encode("utf-8")).hexdigest()[:16]
